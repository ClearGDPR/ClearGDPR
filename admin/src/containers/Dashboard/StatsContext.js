import React, { Component, createContext } from 'react';
import config from 'config';
import { toast } from 'react-toastify';
import PropTypes from 'prop-types';
import internalFetch from 'helpers/internal-fetch';

const StatsContext = createContext({
  isLoading: false,
  stats: {},
  fetchStats: () => {}
});

export class StatsProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.element)
    ])
  };

  state = {
    isLoading: false,
    stats: {
      controller: {},
      processors: []
    },
    fetchStats: this.fetchStats.bind(this)
  };

  request = null;

  setLoading = () => {
    this.setState({ loading: true });
  };

  fetchStats() {
    // prevents doubling up on requests that are in flight
    if (this.state.isLoading) return this.request;
    this.request = internalFetch(`${config.API_URL}/api/management/stats`)
      .then(({ data }) => {
        this.setState({ loading: false, stats: data });
      })
      .catch(err => {
        toast.error(err.message);
        this.setState({ loading: false });
      });
    return this.request;
  }

  render() {
    return <StatsContext.Provider value={this.state}>{this.props.children}</StatsContext.Provider>;
  }
}

export const StatsConsumer = StatsContext.Consumer;
