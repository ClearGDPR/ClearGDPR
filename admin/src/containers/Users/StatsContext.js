import React, { Component, createContext } from 'react';
import config from 'config';
import { toast } from 'react-toastify';
import PropTypes from 'prop-types';
import internalFetch from 'helpers/internal-fetch';
import { PropTypes } from 'victory';

const StatsContext = createContext({
  isLoading: false,
  stats: {},
  fetchStats: () => {}
});

export class StatsProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.element)
    ])
  };

  state = {
    isLoading: false,
    stats: {},
    fetchStats: this.fetchStats
  };

  request = null;

  setLoading = () => {
    this.setState({ loading: true });
  };

  fetchStats = () => {
    // prevents doubling up on requests that are in flight
    if (this.state.isLoading) return this.request;
    this.request = internalFetch(`${config.API_URL}/api/management/stats`)
      .then(({ data: { controller } }) => {
        this.setState({ loading: false, data: controller });
      })
      .catch(err => {
        toast.error(err.message);
        this.setState({ loading: false });
      });
    return this.request;
  };

  render() {
    return <StatsContext.Provider value={this.state}>{this.props.children}</StatsContext.Provider>;
  }
}
