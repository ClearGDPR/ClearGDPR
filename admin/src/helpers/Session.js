import cookie from 'cookie';

export class Session {
  set(authResult) {
    const maxAge = 1 * 24 * 60 * 60; // 1 day long;
    document.cookie = cookie.serialize('auth', authResult, { maxAge });
    document.cookie = cookie.serialize('expires_at', authResult, { maxAge });
  }

  destroy() {
    document.cookie = cookie.serialize('auth', '', { maxAge: -1 });
    document.cookie = cookie.serialize('expires_at', '', { maxAge: -1 });
  }

  isLoggedIn() {
    let expiresAt = cookie.parse(document.cookie).expires_at || '';
    return new Date().getTime() < expiresAt;
  }
}

export default new Session();
