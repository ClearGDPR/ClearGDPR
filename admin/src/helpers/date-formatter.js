import { format } from 'date-fns';

export default date => {
  return format(new Date(date), 'DD/MM/YYYY h:mma');
};
