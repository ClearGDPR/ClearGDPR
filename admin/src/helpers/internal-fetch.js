import merge from 'lodash/object/merge';
import session from '../../helpers/Session';

const internalFetch = (url, config = {}) => {
  const internalConfig = merge(
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${session.getToken()}`,
        'Content-Type': 'application/json'
      }
    },
    config
  );
  return fetch(url, internalConfig).then(async res => {
    if (res.status === 400 && (await res.json()).error === 'JWT token expired') {
      session.destroy();
    } else {
      return res;
    }
  });
};

export default internalFetch;
