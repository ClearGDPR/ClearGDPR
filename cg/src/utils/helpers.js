const Promise = require('bluebird');
const winston = require('winston');

const { VALID_RUN_MODES } = require('./constants');

function inControllerMode() {
  return process.env.MODE === VALID_RUN_MODES.CONTROLLER;
}

function getMyAddress() {
  return process.env.MY_ADDRESS;
}
function getTokenExpiry() {
  return process.env.JWT_TOKEN_EXPIRY || 604800;
}

const timeout = ms => new Promise(res => setTimeout(res, ms));

async function retryAsync(promise, options = {}) {
  let tries = 0;

  const maxRetries = options.maxRetries || 3;
  const interval = options.interval || 500;
  let delay = 0;

  while (tries < maxRetries) {
    await Promise.delay(delay);
    tries++;
    try {
      return await promise();
    } catch (e) {
      delay = interval * tries;
      winston.info(`Attempt no. ${tries} failed. Next attempt in ${delay}ms`);
      winston.log('debug', `Error: ${e.toString()}`);
    }
  }

  winston.error(`Operation failed after ${tries} attempts`);
}

module.exports = {
  getTokenExpiry,
  inControllerMode,
  getMyAddress,
  timeout,
  retryAsync
};
