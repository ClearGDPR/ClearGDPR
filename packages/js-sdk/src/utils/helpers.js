import { merge } from 'lodash/object';

export function parseJSON(response) {
  const contentType = response.headers.get('Content-Type');

  if (contentType.indexOf('application/json') > -1) {
    return new Promise(resolve =>
      response.json().then(json => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: json
        });
      })
    );
  } else {
    return new Promise(resolve =>
      response.text().then(text => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: text
        });
      })
    );
  }
}

export function buildRequestBody(body) {
  return body ? JSON.stringify(body) : null;
}

/**
 * Merges objects values
 * @param {Object} options objects to be merged
 * @returns {Object} New merged object based on params
 */
export function mergeOptions(...objects) {
  return merge(...objects);
}
