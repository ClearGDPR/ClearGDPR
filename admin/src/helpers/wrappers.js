import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';
import { UsersProvider } from 'containers/Users/UsersContext';
import { ProcessorsProvider } from 'containers/Processors/ProcessorsContext';
import { PanelProvider } from 'containers/MainLayout/PanelContext';
import { RectificationsProvider } from 'containers/Rectifications/RectificationsContext';
import { StatsProvider } from 'containers/Dashboard/StatsContext';

export const withWrapper = Wrapper => Component => {
  let wrapper = props => (
    <Wrapper>
      <Component {...props}>{props.children}</Component>
    </Wrapper>
  );
  wrapper.propTypes = {
    children: PropTypes.node
  };
  return wrapper;
};

export const withContextProviders = lodash.flow.apply(null, [
  withWrapper(UsersProvider),
  withWrapper(ProcessorsProvider),
  withWrapper(RectificationsProvider),
  withWrapper(PanelProvider),
  withWrapper(StatsProvider)
]);
