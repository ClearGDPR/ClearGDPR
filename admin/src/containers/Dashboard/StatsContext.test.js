import React from 'react';

import { shallow } from 'enzyme';

import { StatsProvider } from './StatsContext';
import session from 'helpers/Session';
import * as TestUtils from 'tests/helpers/TestUtils';

jest.mock('helpers/Session');

const setupShallow = propOverrides => {
  const props = Object.assign({}, propOverrides);
  const component = shallow(<StatsProvider {...props} />);

  return { props, component };
};

const data = {
  controller: {
    total: 1,
    consented: 1,
    unconsented: 0
  },
  processors: {
    1: {
      consented: 1,
      name: 'test'
    }
  }
};

beforeEach(() => {
  session.getToken.mockReturnValue('token');
});

describe('StatsProvider', () => {
  it('should have correct state after fetching users', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        status: 200,
        json: () => {
          return { data };
        }
      })
    );

    const { component } = setupShallow();
    component.instance().fetchStats();

    await TestUtils.flushPromises();

    expect(component.state()).toEqual(
      expect.objectContaining({
        stats: data,
        isLoading: false
      })
    );
  });

  it('should render correctly', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        status: 200,
        json: () => {
          return { data };
        }
      })
    );

    const { component } = setupShallow();

    await TestUtils.flushPromises();
    expect(component).toMatchSnapshot();
  });
});
