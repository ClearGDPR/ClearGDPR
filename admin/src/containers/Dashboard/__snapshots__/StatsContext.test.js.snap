// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsProvider should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StatsProvider />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": undefined,
      "value": Object {
        "fetchStats": [Function],
        "isLoading": false,
        "stats": Object {
          "controller": Object {},
          "processors": Array [],
        },
      },
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "fetchStats": [Function],
          "isLoading": false,
          "stats": Object {},
        },
        "_currentValue2": Object {
          "fetchStats": [Function],
          "isLoading": false,
          "stats": Object {},
        },
        "_defaultValue": Object {
          "fetchStats": [Function],
          "isLoading": false,
          "stats": Object {},
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": undefined,
        "value": Object {
          "fetchStats": [Function],
          "isLoading": false,
          "stats": Object {
            "controller": Object {},
            "processors": Array [],
          },
        },
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "fetchStats": [Function],
            "isLoading": false,
            "stats": Object {},
          },
          "_currentValue2": Object {
            "fetchStats": [Function],
            "isLoading": false,
            "stats": Object {},
          },
          "_defaultValue": Object {
            "fetchStats": [Function],
            "isLoading": false,
            "stats": Object {},
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
